### PR TITLE
add subcommand summary

### PR DESCRIPTION
**Pull request by Hexilee**
Print statistics of directories in the tree structure.

```bash
> ./juicefs summary -h
NAME:
   juicefs summary - Show tree summary of a directory

USAGE:
   juicefs summary [command options] PATH/INODE

CATEGORY:
   INSPECTOR

DESCRIPTION:
   It is used to show tree summary of target directory.
    
    Examples:
    # Show with path
    $ juicefs summary /mnt/jfs/foo
    
    # Show max depth of 5
    $ juicefs summary --depth 5 /mnt/jfs/foo
   
    # Show top 20 entries
    $ juicefs summary --entries 20 /mnt/jfs/foo
   
    # Show accurate result
    $ juicefs summary --strict /mnt/jfs/foo

OPTIONS:
   --depth value, -d value    depth of tree to show (zero means only show root) (default: 2)
   --entries value, -e value  show top N entries (sort by size) (default: 10)
   --strict                   show accurate summary, including directories and files (may be slow) (default: false)
   --help, -h                 show help (default: false)
```

### Example

```bash
> ./juicefs summary ~/mnt/jfs/
/home/hexi/mnt/jfs/: 490959                       135972.5/s   
/home/hexi/mnt/jfs/: 5.9 GiB (6347730944 Bytes)   1.6 GiB/s    
+--------------------+---------+-------+--------+
|        PATH        |   SIZE  |  DIRS |  FILES |
+--------------------+---------+-------+--------+
| .                  | 5.9 GiB | 47203 | 490959 |
| lani               | 2.8 GiB | 23570 | 236454 |
| lani/common        | 2.8 GiB | 23354 | 235378 |
| lani/apps          | 5.5 MiB |   123 |    840 |
| lani/.git          | 3.3 MiB |    16 |     26 |
| lani/docs          | 880 KiB |    17 |     48 |
| lani/tools         | 500 KiB |    23 |     79 |
| lani/libs          | 464 KiB |    27 |     65 |
| lani/.github       |  24 KiB |     1 |      4 |
| lani/rush.json     |  24 KiB |     0 |      1 |
| lani/node_modules  | 8.0 KiB |     0 |      2 |
| lani/.vscode       | 8.0 KiB |     0 |      2 |
| lani/...           |  36 KiB |     0 |      9 |
| lani1              | 2.8 GiB | 23570 | 236454 |
| lani1/common       | 2.8 GiB | 23354 | 235378 |
| lani1/apps         | 5.5 MiB |   123 |    840 |
| lani1/.git         | 3.3 MiB |    16 |     26 |
| lani1/docs         | 880 KiB |    17 |     48 |
| lani1/tools        | 500 KiB |    23 |     79 |
| lani1/libs         | 464 KiB |    27 |     65 |
| lani1/.github      |  24 KiB |     1 |      4 |
| lani1/rush.json    |  24 KiB |     0 |      1 |
| lani1/.vscode      | 8.0 KiB |     0 |      2 |
| lani1/node_modules | 8.0 KiB |     0 |      2 |
| lani1/...          |  36 KiB |     0 |      9 |
| test               | 332 MiB |     0 |  17912 |
| test/file14698.bin |  36 KiB |     0 |      1 |
| test/file6303.bin  |  36 KiB |     0 |      1 |
| test/file14862.bin |  36 KiB |     0 |      1 |
| test/file17670.bin |  36 KiB |     0 |      1 |
| test/file050.bin   |  36 KiB |     0 |      1 |
| test/file099.bin   |  36 KiB |     0 |      1 |
| test/file1943.bin  |  36 KiB |     0 |      1 |
| test/file12992.bin |  36 KiB |     0 |      1 |
| test/file4031.bin  |  36 KiB |     0 |      1 |
| test/file4792.bin  |  36 KiB |     0 |      1 |
| test/...           | 332 MiB |     0 |  17902 |
| roa                | 2.2 MiB |    58 |    139 |
| roa/.git           | 1.4 MiB |    16 |     26 |
| roa/roa            | 248 KiB |     8 |     30 |
| roa/integration    | 144 KiB |    12 |     22 |
| roa/roa-core       | 120 KiB |     3 |     17 |
| roa/roa-async-std  |  32 KiB |     1 |      6 |
| roa/.github        |  28 KiB |     1 |      6 |
| roa/roa-diesel     |  28 KiB |     1 |      5 |
| roa/examples       |  28 KiB |     0 |      7 |
| roa/assets         |  24 KiB |     1 |      5 |
| roa/.vscode        |  20 KiB |     0 |      1 |
| roa/...            |  68 KiB |     1 |     14 |
+--------------------+---------+-------+--------+
```

### CODE REVIEW

File: pkg/vfs/internal.go [hunk 1]
```diff
@@ -302,6 +306,36 @@ func (r *InfoResponse) Decode(reader io.Reader) error {
...
+	buffer := utils.NewBuffer(4 + uint32(len(resp)))
+	buffer.Put32(uint32(len(resp)))
+	buffer.Put(resp)
+	return buffer.Bytes()
+}
+
+func (r *SummaryReponse) Decode(reader io.Reader) error {
+	sizeBuf := make([]byte, 4)
+	n := 0
+	for n == 0 {
```
Comment (davies) [inline pkg/vfs/internal.go:320]: n < 4

File: pkg/vfs/internal.go [hunk 2]
```diff
@@ -464,6 +498,39 @@ func (v *VFS) handleInternalMsg(ctx meta.Context, cmd uint32, r *utils.Buffer, o
...
+		}
+
+		done := make(chan struct{})
+		var r syscall.Errno
+		go func() {
+			r = v.Meta.GetTreeSummary(ctx, &tree, depth, topN, dirOnly)
+			close(done)
+		}()
+		writeProgress(&tree.Files, &tree.Size, out, done)
+		_, _ = out.Write([]byte{uint8(r)})
```
Comment (davies) [inline pkg/vfs/internal.go:528]: after waiting below

File: cmd/summary.go [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+			t = 100
+		}
+		topN = uint8(t)
+	}
+	if ctx.Bool("strict") {
+		dirOnly = 0
+	}
+
+	progress := utils.NewProgress(false)
+	for i := 0; i < ctx.Args().Len(); i++ {
```
Comment (davies) [inline cmd/summary.go:118]: only one PATH

File: cmd/summary.go [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+ # Show top 20 entries
+ $ juicefs summary -t 20 /mnt/jfs/foo
+
+ # Show accurate result
+ $ juicefs summary --strict /mnt/jfs/foo
+ `,
+		Flags: []cli.Flag{
+			&cli.BoolFlag{
+				Name:    "inode",
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply cmd/summary.go]: Removed

File: pkg/vfs/internal.go (hunk 2 above)
Comment (Hexilee) [reply pkg/vfs/internal.go]: Fixed, `readProgress` now support read data directly.

File: pkg/vfs/internal.go [hunk 5]
```diff
@@ -464,6 +470,45 @@ func (v *VFS) handleInternalMsg(ctx meta.Context, cmd uint32, r *utils.Buffer, o
...
+
+		done := make(chan struct{})
+		var r syscall.Errno
+		go func() {
+			r = v.Meta.GetTreeSummary(ctx, &tree, depth, topN, strict)
+			close(done)
+		}()
+		writeProgress(&tree.Files, &tree.Size, out, done)
+		resp := &SummaryReponse{Tree: tree}
+		if r != 0 {
```
Comment (davies) [inline pkg/vfs/internal.go:501]: not needed

File: pkg/meta/interface.go [hunk 6]
```diff
@@ -230,6 +233,18 @@ type Summary struct {
 	Dirs   uint64
 }
 
+type TreeSummary struct {
+	Inode    Ino
+	Path     string
+	Type     uint8
```
Comment (davies) [inline pkg/meta/interface.go:239]: not used

File: pkg/meta/utils.go [hunk 7]
```diff
@@ -476,3 +477,239 @@ func (m *baseMeta) FastGetSummary(ctx Context, inode Ino, summary *Summary, recu
...
+				*stat = *s
+				var attr Attr
+				if st := m.GetAttr(ctx, tree.Inode, &attr); st != 0 && st != syscall.ENOENT {
+					return st
+				}
+				if attr.Nlink == 2 {
+					// leaf dir, no need to read entries
+					return nil
+				}
+				if st := m.Readdir(ctx, tree.Inode, 0, entries); st != 0 && st != syscall.ENOENT {
```
Comment (davies) [inline pkg/meta/utils.go:670]: use m.doReaddir() to skip `.` and `..`

File: pkg/meta/utils.go [hunk 8]
```diff
@@ -476,3 +477,239 @@ func (m *baseMeta) FastGetSummary(ctx Context, inode Ino, summary *Summary, recu
...
+func (t *TreeSummary) visitRoot(visit func(*TreeSummary)) {
+	for tree := t; tree != nil; tree = tree.parent {
+		visit(tree)
+	}
+}
+
+func (m *baseMeta) GetTreeSummary(ctx Context, root *TreeSummary, depth, topN uint8, strict bool) syscall.Errno {
+	var st syscall.Errno
+	if strict {
+		st = m.getTreeSummary(ctx, root)
```
Comment (davies) [inline pkg/meta/utils.go:490]: We don't need to details of tree beyond `depth`, use GetSummary for them.

it's better to merge this two together

File: cmd/summary.go [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+	if runtime.GOOS == "windows" {
+		logger.Infof("Windows is not supported")
+		return nil
+	}
+	var depth, topN, strict uint8
+	depth = 2
+	topN = 10
+	strict = 0
+
+	if ctx.IsSet("depth") {
```
Comment (davies) [inline cmd/summary.go:94]: depth := int(ctx.Uint("depth")

File: pkg/vfs/internal.go (hunk 5 above)
Comment (Hexilee) [reply pkg/vfs/internal.go]: Removed

File: pkg/meta/utils.go [hunk 10]
```diff
@@ -476,3 +477,239 @@ func (m *baseMeta) FastGetSummary(ctx Context, inode Ino, summary *Summary, recu
...
+				s, st := m.GetDirStat(ctx, tree.Inode)
+				if st != 0 {
+					return st
+				}
+				*stat = *s
+				var attr Attr
+				if st := m.GetAttr(ctx, tree.Inode, &attr); st != 0 && st != syscall.ENOENT {
+					return st
+				}
+				if attr.Nlink == 2 {
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/meta/utils.go]: Done

File: pkg/meta/utils.go (hunk 7 above)
Comment (Hexilee) [reply pkg/meta/utils.go]: Replaced

File: cmd/summary.go [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+	"github.com/urfave/cli/v2"
+)
+
+func cmdSummary() *cli.Command {
+	return &cli.Command{
+		Name:      "summary",
+		Action:    summary,
+		Category:  "INSPECTOR",
+		Usage:     "Show tree summary of a directory",
+		ArgsUsage: "PATH/INODE",
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply cmd/summary.go]: Fixed

File: cmd/summary.go (hunk 9 above)
Comment (Hexilee) [reply cmd/summary.go]: Done

File: pkg/meta/utils.go (hunk 8 above)
Comment (Hexilee) [reply pkg/meta/utils.go]: Done

**Pull request merged**